browser/ui/webui/brave_rewards/tip_panel_ui.cc

```diff
@@ -0,0 +1,96 @@
+
+namespace brave_rewards {
+
+namespace {
+
+static constexpr webui::LocalizedString kStrings[] = {
+    {"balanceLabel", IDS_REWARDS_TIP_BALANCE_LABEL},
+    {"monthlyToggleLabel", IDS_REWARDS_TIP_MONTHLY_TOGGLE_LABEL},
+    {"feeNotice", IDS_REWARDS_TIP_FEE_NOTICE},
+    {"termsOfService", IDS_BRAVE_REWARDS_ONBOARDING_TERMS},
+    {"sendFormTitle", IDS_REWARDS_TIP_SEND_FORM_TITLE},
+    {"sendButtonLabel", IDS_REWARDS_TIP_SEND_BUTTON_LABEL},
+    {"web3ButtonLabel", IDS_REWARDS_TIP_WEB3_BUTTON_LABEL},
+    {"verifiedTooltipTitle", IDS_REWARDS_TIP_VERIFIED_TOOLTIP_TITLE},
+    {"verifiedTooltipText", IDS_REWARDS_TIP_VERIFIED_TOOLTIP_TEXT},
+    {"monthlyTooltipText", IDS_REWARDS_TIP_MONTHLY_TOOLTIP_TEXT},
+    {"learnMoreLabel", IDS_REWARDS_TIP_LEARN_MORE_LABEL},
+    {"customAmountLabel", IDS_REWARDS_TIP_CUSTOM_AMOUNT_LABEL},
+    {"monthlySetTitle", IDS_REWARDS_TIP_MONTHLY_SET_TITLE},
+    {"monthlySetText", IDS_REWARDS_TIP_MONTHLY_SET_TEXT},
+    {"providerMismatchTitle", IDS_REWARDS_TIP_PROVIDER_MISMATCH_TITLE},
+    {"providerMismatchText", IDS_REWARDS_TIP_PROVIDER_MISMATCH_TEXT},
+    {"providerMismatchWeb3Text", IDS_REWARDS_TIP_PROVIDER_MISMATCH_WEB3_TEXT},
+    {"web3OnlyTitle", IDS_REWARDS_TIP_WEB3_ONLY_TITLE},
+    {"reconnectTitle", IDS_REWARDS_TIP_RECONNECT_TITLE},
+    {"reconnectText", IDS_REWARDS_TIP_RECONNECT_TEXT},
+    {"reconnectWeb3Text", IDS_REWARDS_TIP_RECONNECT_WEB3_TEXT},
+    {"reconnectButtonLabel", IDS_REWARDS_TIP_RECONNECT_BUTTON_LABEL},
+    {"insufficientBalanceTitle", IDS_REWARDS_TIP_INSUFFICIENT_BALANCE_TITLE},
+    {"insufficientBalanceText", IDS_REWARDS_TIP_INSUFFICIENT_BALANCE_TEXT},
+    {"contributionFailedTitle", IDS_REWARDS_TIP_CONTRIBUTION_FAILED_TITLE},
+    {"contributionFailedText", IDS_REWARDS_TIP_CONTRIBUTION_FAILED_TEXT},
+    {"tryAgainButtonLabel", IDS_REWARDS_TIP_TRY_AGAIN_BUTTON_LABEL},
+    {"contributionSentTitle", IDS_REWARDS_TIP_CONTRIBUTION_SENT_TITLE},
+    {"contributionSentText", IDS_REWARDS_TIP_CONTRIBUTION_SENT_TEXT},
+    {"shareButtonLabel", IDS_REWARDS_TIP_SHARE_BUTTON_LABEL},
+    {"shareText", IDS_REWARDS_TIP_SHARE_TEXT},
+    {"unexpectedErrorTitle", IDS_REWARDS_TIP_UNEXPECTED_ERROR_TITLE},
+    {"unexpectedErrorText", IDS_REWARDS_TIP_UNEXPECTED_ERROR_TEXT},
+    {"defaultCreatorDescription", IDS_REWARDS_TIP_DEFAULT_CREATOR_DESCRIPTION}};
+
+}  // namespace
+
+TipPanelUI::TipPanelUI(content::WebUI* web_ui)
+    : MojoBubbleWebUIController(web_ui, true) {
+  auto* source = content::WebUIDataSource::CreateAndAdd(
+      web_ui->GetWebContents()->GetBrowserContext(), kBraveTipPanelHost);
+
+  source->AddLocalizedStrings(kStrings);
+
+  webui::SetupWebUIDataSource(
+      source, base::make_span(kTipPanelGenerated, kTipPanelGeneratedSize),
+      IDR_TIP_PANEL_HTML);
+}
+
+TipPanelUI::~TipPanelUI() = default;
+
+WEB_UI_CONTROLLER_TYPE_IMPL(TipPanelUI)
+
+void TipPanelUI::BindInterface(
+    mojo::PendingReceiver<TipPanelHandlerFactory> receiver) {
+  factory_receiver_.reset();
```

Comment:
Shouldn't this be `DCHECK(!factory_receiver_.is_bound());` instead?

Reply:
Maybe? This pattern seems to be common in both [Chromium](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/webui/access_code_cast/access_code_cast_ui.cc;l=98?q=%22factory_receiver_.Bind%22&ss=chromium) and Brave, though.

Reply:
You can find both in the codebase. I'm not sure if it's intended to call `BindInterface()` multiple times. If it's not, then it should be a `DCHECK`, since now we're technically unbinding the receiver silently. I'll leave that up to you, though!